consistent-sampling/src/main/java/io/opentelemetry/contrib/export/ConsistentReservoirSamplingBatchSpanProcessor.java

```diff
@@ -0,0 +1,677 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.contrib.export;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.metrics.LongCounter;
+import io.opentelemetry.api.metrics.Meter;
+import io.opentelemetry.api.metrics.MeterProvider;
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.api.trace.TraceState;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.contrib.state.OtelTraceState;
+import io.opentelemetry.contrib.util.RandomGenerator;
+import io.opentelemetry.contrib.util.RandomUtil;
+import io.opentelemetry.sdk.common.CompletableResultCode;
+import io.opentelemetry.sdk.internal.DaemonThreadFactory;
+import io.opentelemetry.sdk.trace.ReadWriteSpan;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+import io.opentelemetry.sdk.trace.SpanProcessor;
+import io.opentelemetry.sdk.trace.data.DelegatingSpanData;
+import io.opentelemetry.sdk.trace.data.SpanData;
+import io.opentelemetry.sdk.trace.export.SpanExporter;
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.Collections;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * Implementation of the {@link SpanProcessor} that batches spans exported by the SDK then pushes
+ * them to the exporter pipeline.
+ *
+ * <p>All spans reported by the SDK implementation are first added to a synchronized queue (with a
+ * {@code maxQueueSize} maximum size, if queue is full spans are dropped). Spans are exported either
+ * when there are {@code maxExportBatchSize} pending spans or {@code scheduleDelayNanos} has passed
+ * since the last export finished.
+ */
+public final class ConsistentReservoirSamplingBatchSpanProcessor implements SpanProcessor {
```

Comment:
Is this for tail sampling? It's not intuitive why we'd have a span processor doing sampling otherwise.

If possible I'd recommend holding off on this class until it's in the spec as documentation there would help a lot with understanding it, intuitively it seems unrelated to `ConsistentSampler`

Reply:
> Is this for tail sampling? It's not intuitive why we'd have a span processor doing sampling otherwise.

No, this is different from tail sampling, where a sampling decision is made for the entire trace after all spans of a trace have been sampled. The ConsistentReservoirSamplingBatchSpanProcessor can be used to implement rate-limited sampling of spans with a hard limit. Since the sampler interface requires an immediate sampling decision, it is not straightforward to ensure that the rate limit is never exceeded, if the number of spans suddenly increases. For example, the ConsistentRateLimitingSampler may exceed the desired span rate in such cases. By delaying the sampling decision, as the ConsistentReservoirSamplingBatchSpanProcessor does, a hard limit can be achieved while giving all spans a fair chance to be sampled.

Reply:
> If possible I'd recommend holding off on this class until it's in the spec as documentation there would help a lot with understanding it, intuitively it seems unrelated to ConsistentSampler

I have removed the ConsistentReservoirSamplingBatchSpanProcessor from this PR (see dd3df4ed697e968ee822228a92ad5aba58c32a6a)